Validate Puppet Version

dhollinger opened the pull request:
closes #99

ghoneycutt commented on .travis.yml:
```diff
@@ -7,10 +7,9 @@ script:
   - 2.1.9
-  - 2.3.1
-  - 2.4.1
+  - 2.4.4
```
Why switch away from ruby versions that Puppet uses? I get dropping the old versions, though this should use 2.4.3 to match the latest Puppet 5 release.

https://puppet.com/docs/puppet/5.5/about_agent.html

dhollinger replied:
I can add 2.4.3 in, I added 2.4.4 based on what PDK uses. The official PDK installation uses Ruby 2.4.4 with Puppet 5.5.1

ghoneycutt commented:
This project is meant to validate that the metadata is correct for a puppet module and the metadata is not invalid because it uses an unsupported version of puppet. I'm against adding this functionality at all. If it is added, there needs to be a simple way to disable it.

bastelfreak commented:
@ghoneycutt this PR is based on a discussion we had yesterday in #voxpupuli. The idea was to check if the supported puppet version range is within the range that Puppetlabs supports.

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +136,26 @@ def validate_requirements!(requirements)
+  def validate_puppet_ver!(range)
+    min_ver = range['version_requirement'].gsub(/[><=~^]/, '').split(' ')[0]
+
+    error(:requirements, "#{min_ver} is no longer a supported version of Puppet. Please use a valid range between 4.10.0 and 6.0.0.") if min_ver =~ /^[2-4]\.[0-9]\.\d$/
```
Module authors should be free to support older versions. I get some people don't want to support EOL versions, but in practice sometimes you have to. Making this required will mean some users will stop using this. We have branches that support older versions and often do cherry picks. It's nice to know the metadata is correct, but if this is required I'll just have to rip the validation from our CI. Please make this configurable.

Mentioning the upper bound is also a bad idea, especially since you don't check for it anyway.

I'd also use `VersionRequirement` instead of parsing it yourself.

ghoneycutt replied:
@dhollinger I think there is confusion between schema and data with regards to what is "current stable". I agree that we should target the "current stable" version of the schema. This is different from the data.

If you have an up to date and totally valid metadata.json and it supports an EOL version of puppet, that should not throw an error by default. That violates the nature of the tool, which is to ensure the metadata.json follows the schema, as well as the principal of least surprise.

rnelson0 replied:
An old version of puppet, even an unsupported one, in `metadata.json` is just part of linting, there's no need for or implication that the testing framework to support that version. That's up to the user.

I suggest using `--strict-supported-versions` and a `if options[:strict_supported_versions]` guard around the error, similar to [this](https://github.com/voxpupuli/metadata-json-lint/blob/39d756d015ef6e6ff75f75af06e5e5c0c1cf3bb7/lib/metadata_json_lint.rb#L106).

bastelfreak commented:
Making this a configureable feature is a good idea, thanks @ghoneycutt. On Vox Pupuli we agreed on only supporting the current Puppet versions that also Puppetlabs support, so a check for that would be helpful. This gem was the first that came to our mind as we thought about a good place to implement it.

ghoneycutt commented:
@dhollinger I'm actually all for this functionality and would use it if it was in a separate project since it is not related to verifying if the metadata is correct.

dhollinger commented on tests/requirements_bad_version/Rakefile:
```diff
@@ -0,0 +1,2 @@
+$LOAD_PATH.unshift(File.expand_path('../../../lib', __FILE__))
```
I just copied the Rakefile from the other tests

ghoneycutt replied:
My comment was about the Rakefile but was about the naming. `bad version` implies that the version is not properly formatted for the schema. This should use wording such as `old` or `eol`.

dhollinger commented:
I'll work on adding the configurable option later today/tonight

ghoneycutt commented on tests/requirements_bad_version/expected:
```diff
@@ -0,0 +1 @@
+(ERROR) requirements: 3.2.0 is no longer a supported version of Puppet. Please use a valid range between 4.10.0 and 6.0.0.
```
Should probably not specify an upper bound and just check the lower bound.

ekohl commented:
> I'll work on adding the configurable option later today/tonight

How about a setting that accepts the minimum version so users can set it to something like `4.8`. It should also support something like `current`. That should detect the current version. It probably means we also need to prepare it for future EOL dates.

dhollinger commented:
@ekohl @rnelson0 I'm all for making it configurable, but as a validation tool, I don't think we should allow the lower bound limit to be set. It sets a precedent that we will support options for EOL/Unsupported versions of Puppet. I think it should be a simple on/off based option like the `--strict-supported-versions` suggested option. 

I'll do whatever the community as a whole agrees on, but we don't have the man power to support every potential option a user needs as the larger user base as a whole still has Puppet 2 users.

rnelson0 commented:
This isn't about our support, it's about users validating their own modules. What "we" support would be defined in the default, but if I am stuck at 4.2 at work, I should be able to specify that. It won't reflect on VP at all.

dhollinger commented:
I guess the argument I'm trying to make is that testing tools should not account for every potential option. If it did, the `rspec`, `rspec-puppet`, `cucumber`, etc would have to support options for every version of Puppet/RSpec/Ruby ever because a user might still be using an old version. We should not support/enable that, instead it should be a simply on/off option, or, if they want to continue using an older version of software, they have to use the older versions of tools that go with that.

ekohl commented:
I agree with @rnelson0: I see metadata-json-lint as a generic tool, not a voxpupuli tool. Those settings are meant for users while voxpupuli will probably just stick with supported versions.

dhollinger commented:
@rnelson0 @ekohl @ghoneycutt 

In relation to the "problem" this PR is trying to solve, I think that for the initial roll out of something like this should focus on being simple initially. So, add the functionality to validate `strict-requirements` for puppet versions as default with the ability to disable it.

We can return to the discussion of adding the ability to add a lower (and/or upper) bound later as a separate PR or as the start of a plugin system for metadata lint as per @llowder's suggestion.

Thoughts?

@bastelfreak @binford2k - anything to add?

binford2k commented:
Nothing to add, but this has been a fascinating conversation to watch :D

bastelfreak commented:
@dhollinger I agree with your and @binford2ks last statement.

dhollinger commented:
@ghoneycutt @ekohl @rnelson0 - I've updated the PR with a few changes based on the discussion:
* Added a `--[no-]strict-puppet-version` option to the gem and set it to default to false.
* Updated the output when the `strict-puppet-version` is turned on it now outputs a WARN instead of an ERROR.
* Renamed the related test from `requirements_bad_version` to `requirements_eol_version`.
* Added Ruby `2.4.3` back into the travis matrix since that ships with the more recent Puppet Agent packages. Kept `2.4.4` since that is what the Puppet Development Kit will test on by default.

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +142,26 @@ def validate_requirements!(requirements)
   module_function :validate_requirements!
 
+  def validate_puppet_ver!(range)
+    min_ver = range['version_requirement'].gsub(/[><=~^]/, '').split(' ')[0]
```
Can we parse this using `VersionRequirement` rather than a regex? This will also fail with a nil error if the key is not set.

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +142,26 @@ def validate_requirements!(requirements)
+        error :dependencies, "Invalid 'version_requirement' field in metadata.json: #{e}"
+      end
+
+      validate_puppet_ver!(requirement) if options[:strict_puppet_version]
```
Here you can use the parsed `VersionRequirement` rather than the raw requirement. If it failed to parse, validating the version is likely to fail and I wouldn't bother validating the content.

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +142,26 @@ def validate_requirements!(requirements)
+  def validate_puppet_ver!(range)
+    min_ver = range['version_requirement'].gsub(/[><=~^]/, '').split(' ')[0]
+
+    warn(:requirements, "#{min_ver} is no longer a supported version of Puppet. Minimum supported version is 4.10.0.") if min_ver =~ /^[2-4]\.[0-9]\.\d$/
```
`3.x` is semantically valid but you'd miss it with the current code. Same for `>= 3.0` or if anyone used `3.10` instead of `4.10`

dhollinger replied:
I'll take a look at this and fix it.

dhollinger replied:
@ekohl scratch that. I am going to update the code to use the VersionRequirement class instead of a bunch of duplicate inline code, but we already warn on the use of things like `3.x` and `3.*`, which by default will cause a failure as `--fail-on-warnings` is turned on by default,

dhollinger commented:
@ekohl I will work on having the `requirements` section of the `metadata.json` file use the existing ruby class(es) to parse and validate the semver of the requirements section. I think that there should either be a `strict-puppet-semver` option that enforces proper semver format, or it should always be enforced as the puppet version will _always_ be in semver format unless a user is building their own puppet (which at that point they're probably not using this gem).

As for minimum version, I still think that should be on/off only (defaulted to off). Hard-coding the minimum puppet version is no different than what other gems/libraries would do for their validation and just gets updated as that matrix shifts. I highly doubt very many users that use this gem will care about minimum version beyond what the "supported" it or isn't. Everything else is an edge case and it is not a good idea in any way for us to try support a bunch of edge cases as they are just that, outside of what the norm is.

I think @llowder's suggestion is better for that as it would allow VP or users to create their own plugins to add functionality for such edge cases. I treat this as I would Chef's `metadata.rb` or RubyGems `*.gemspec` files. It's validating a file descriptive of the codebase and linting should be checked or enforced based on how the tools (like `puppet module install`, `puppet forge`, and/or `r10k`) utilize that file rather than based on potential edge cases.

ekohl commented:
> I think @llowder's suggestion is better for that as it would allow VP or users to create their own plugins to add functionality for such edge cases. I treat this as I would Chef's metadata.rb or RubyGems *.gemspec files. It's validating a file descriptive of the codebase and linting should be checked or enforced based on how the tools (like puppet module install, puppet forge, and/or r10k) utilize that file rather than based on potential edge cases.

I agree very much with this: if `metadata-json-lint` guarantees tools like puppet, forge and r10k accept your metadata as you intended it then that's a great scope. Opinions about content are controversial which makes it hard to reach a consensus as can be seen by this PR. Plugins can be a great solution.

rnelson0 commented on lib/metadata_json_lint.rb:
```diff
@@ -40,6 +42,10 @@ def run
       end
 
+      opts.on('--[no-]strict-puppet-version', "Fail on strict Puppet Version check based on current supported Puppet versions. Defaults to '#{options[:strict_puppet_version]}'.") do |v|
+        options[:strict_puppet_version] = v
```
Given some concerns about changing option names and moving this to a plugin, perhaps this should be marked `[Experimental]` or with similar text?

dhollinger replied:
I don't think we're planning on moving this option specifically to a plugin. This is designed to be the puppet version equivalent to `--strict-dependencies`. Simply enforce puppet version requirements based on supported versions and semver requirements. An on/off option that, like `--strict-dependencies` is off by default.

Additional tests, like being able to specify your own version minimum and test against that, would be a plugin as it's outside the simple scope of the base tool.

rnelson0 commented on tests/duplicate-dep/metadata.json:
```diff
@@ -57,13 +57,9 @@
     {
       "name": "puppet",
-      "version_requirement": "3.x"
+      "version_requirement": "4.10.0"
```
To prevent having to change this constantly as the minimum supported version is updated, I suggest setting it to the current version, `5.5.1`.

dhollinger commented:
@ekohl I'd like this PR to simply be for a simple on/off switch for enforcing strict puppet versions (Supported version + proper SemVer) only and is defaulted to off. Any additional functionality I'd like to push to a discussion about a plugins system for metadata.json.

Thoughts?

dhollinger commented:
Squashed commits into single commit

ekohl commented on lib/metadata-json-lint/version_requirement.rb:
```diff
@@ -47,6 +47,22 @@ def open_ended?
     end
 
+    def puppet_eol?
+      true if range.begin.to_s =~ /^[2-4]\.[0-9]\.\d$/
```
I really don't like a regex for version comparisons. Isn't there a real function to do comparison?

dhollinger replied:
To my knowledge there is not a function for comparing SemVer strings. Not that I've found anyway. Every lookup I've done on it ends up in even more complex loops and regex.

ekohl replied:
`range.begin < SemanticPuppet::Version.parse(MIN_PUPPET_VER)` would work. In my test I actually moved the constant to `MetadataJsonLint::MIN_PUPPET_VER` so it wasn't a constant in the global scope.

ekohl commented on README.md:
```diff
@@ -76,6 +61,7 @@ MetadataJsonLint.options.strict_license = false
 * `--[no-]strict-dependencies`: Whether to fail if module version dependencies are open-ended. Defaults to `false`.
 * `--[no-]strict-license`: Whether to fail on strict license check. Defaults to `true`.
 * `--[no-]fail-on-warnings`: Whether to fail on warnings. Defaults to `true`.
+* `--[no-]strict-puppet-version`: Whether to fail if Puppet version requirements are open-ended and no longer supported. Defaults to `false`.
```
Doesn't open-ended mean there aren't limits one side? Should this be `open-ended or no longer`?

dhollinger replied:
It should be `open-ended or no longer supported`. I'll double check the code, but I'm pretty sure the `open-ended` check is specifically checking for an open-ended upper bound. It doesn't make much sense to check for open endedness on the lower bound. 

Or rather, I've never seen version requirements that have no minimum version by do have a maximum version (though I'm sure they exist somewhere)

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +142,34 @@ def validate_requirements!(requirements)
+        puppet_req = VersionRequirement.new(requirement.fetch('version_requirement', ''))
+      rescue ArgumentError => e
+        # Raised when the version_requirement provided could not be parsed
+        error :dependencies, "Invalid 'version_requirement' field in metadata.json: #{e}"
```
Does this raise an exception or will the execution continue?

dhollinger replied:
That should raise an error in the output (using the error method) detailing that the version is not SemVer compatible. I put this in there as it was not being checked at all and because you can't use a non-SemVer version for Puppet ranges. This is the same check as is in the `validate_dependencies!` method.

ekohl replied:
My underlying question is actually: does it proceed with `validate_puppet_ver!(puppet_req)` (which would fail).

ekohl replied:
Could we add a test for this? If it doesn't crash then I'm fine with it but I suspect this will crash on a `NameError: undefined local variable or method`

dhollinger replied:
@ekohl There is already a test for it found here: https://github.com/voxpupuli/metadata-json-lint/blob/master/spec/version_requirement_spec.rb#L5

ekohl replied:
What I meant for this was https://gist.github.com/ekohl/f86924dc00197890a831d57fd6329324:

```
Failures:

  1) MetadataJsonLint.validate_requirements! with invalid requirement should not raise Exception
     Failure/Error: expect { described_class.validate_requirements!(requirements) }.not_to raise_error
     
       expected no Exception, got #<NoMethodError: undefined method `mixed_syntax?' for nil:NilClass> with backtrace:
         # ./lib/metadata_json_lint.rb:169:in `validate_puppet_ver!'
         # ./lib/metadata_json_lint.rb:155:in `block in validate_requirements!'
         # ./lib/metadata_json_lint.rb:143:in `each'
         # ./lib/metadata_json_lint.rb:143:in `validate_requirements!'
         # ./spec/metadata_json_lint_spec.rb:33:in `block (5 levels) in <top (required)>'
         # ./spec/metadata_json_lint_spec.rb:33:in `block (4 levels) in <top (required)>'
     # ./spec/metadata_json_lint_spec.rb:33:in `block (4 levels) in <top (required)>'
```

The alternative is an example metadata.json.

Looks like it can be mitigated by adding an `if puppet_req` to the line below.

Also wondering if the error here should be `:dependencies`. Looking at the rest it looks like it should be `:requirements`.

ekohl replied:
> Also wondering if the error here should be :dependencies. Looking at the rest it looks like it should be :requirements.

Could you have a look at this?

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +142,34 @@ def validate_requirements!(requirements)
+    end
+
+    if options[:strict_puppet_version] && requirement.puppet_eol?
+      warn(:requirements, "#{requirement.min} is no longer supported. Minimum supported version is 4.10.0")
```
I really don't like hardcoded versions. It's just waiting to be outdated. It should at least be a constant somewhere.

ekohl commented:
> @ekohl I'd like this PR to simply be for a simple on/off switch for enforcing strict puppet versions (Supported version + proper SemVer) only and is defaulted to off. Any additional functionality I'd like to push to a discussion about a plugins system for metadata.json.

I don't like building in technical debt from the start. Hardcoding values is a bad engineering practice. They should at least be defined as constants at the top of the class. That means it's very easy for the maintainer to change the values, even if the user can't just now.

dhollinger commented:
@ekohl I can agree with that. Being in a constant makes it easier to find when it needs to be changed as well.

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -8,17 +8,21 @@
 require 'metadata-json-lint/schema'
 require 'metadata-json-lint/version_requirement'
 
+MIN_PUPPET_VER = '4.10.0'.freeze
```
I think it's cleaner to move this inside `module MetadataJsonLint` so it's not in the global scope.

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +144,34 @@ def validate_requirements!(requirements)
+      warn(:requirements, "#{requirement.min} is no longer supported. Minimum supported version is #{MIN_PUPPET_VER}")
+    end
+
+    return unless requirement.mixed_syntax?
```
Nit: any reason to use a return here instead of another if statement here? This feels a bit unnatural to me but I don't really care that much.

ekohl commented on spec/metadata_json_lint_spec.rb:
```diff
@@ -0,0 +1,37 @@
+      end
+    end
+
+    # context 'with invalid requirement' do
```
Any reason this is commented?

ekohl replied:
And a look at this?

dhollinger commented:
@ekohl any other comments or concerns?

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +144,34 @@ def validate_requirements!(requirements)
+    end
+
+    return unless requirement.mixed_syntax?
+    warn(:dependencies, 'Mixing "x" or "*" version syntax with operators is not recommended in ' \
```
`:dependencies` here should probably be `:requirements`

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +144,34 @@ def validate_requirements!(requirements)
+
+    return unless requirement.mixed_syntax?
+    warn(:dependencies, 'Mixing "x" or "*" version syntax with operators is not recommended in ' \
+      "metadata.json, use one style in the #{dep['name']} dependency: #{dep['version_requirement']}")
```
`dep` is probably not defined here and it should look at `requirement` instead.

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -167,8 +167,8 @@ def validate_puppet_ver!(requirement)
-    warn(:dependencies, 'Mixing "x" or "*" version syntax with operators is not recommended in ' \
-      "metadata.json, use one style in the #{dep['name']} dependency: #{dep['version_requirement']}")
+    warn(:requirements, 'Mixing "x" or "*" version syntax with operators is not recommended in ' \
+      "metadata.json, use one style in the #{requirements['name']} dependency: #{requirements['version_requirement']}")
```
This only has the variable `requirement` in the local scope. The name isn't known, but we do know that this is the puppet requirement. Perhaps `VersionRequirement` needs a `.to_s` for this.

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +144,34 @@ def validate_requirements!(requirements)
+        error :requirements, "Invalid 'version_requirement' field in metadata.json: #{e}"
+      end
+
+      validate_puppet_ver!(puppet_req) unless puppet_req.instance_variable_get('@requirement').nil?
```
If we use this variable, perhaps it should be a public property?

ekohl commented on lib/metadata_json_lint.rb:
```diff
@@ -136,10 +144,34 @@ def validate_requirements!(requirements)
+
+    return unless requirement.mixed_syntax?
+    warn(:requirement, 'Mixing "x" or "*" version syntax with operators is not recommended in ' \
+      "metadata.json, use one style in the puppet version: #{requirement.instance_variable_get('@requirement')}")
```
Same here
